R/utils.R

```diff
@@ -259,85 +259,560 @@ rowcol_to_dim <- function(row, col) {
+    stop("only `rows` and `cols` can be provided without names. You must name all other arguments.")
+  }
+  if (lengt == 1 && all_args_unnamed) {
+    stop(
+      "Supplying a single unnamed argument is not handled by `wb_dims()`",
+      "use `x`, `from_row` / `from_col`. You can also use `dims = NULL`"
+    )
+  }
+  ok_if_arg1_unnamed <-
+    is.atomic(args[[1]]) | any(nams %in% c("rows", "cols"))
 
-  scol_null <- is.null(scol)
-  srow_null <- is.null(srow)
+  if (nams[1] == "" && !ok_if_arg1_unnamed) {
+    stop(
+      "The first argument must either be named or be a vector.",
+      "Providing a single named argument must either be `from_row`, `from_col` or `x`."
+    )
+  }
+  if (n_unnamed_args == 1 && lengt > 1 && !"rows" %in% nams) {
+    message("Assuming the first unnamed argument to be `rows`.")
+    nams[which(nams == "")[1]] <- "rows"
+    names(args) <- nams
+    n_unnamed_args <- length(which(!nzchar(nams)))
+    all_args_unnamed <- n_unnamed_args == lengt
+  }
+  if (n_unnamed_args == 1 && lengt > 1 && "rows" %in% nams) {
+    message("Assuming the first unnamed argument to be `cols`.")
+    nams[which(nams == "")[1]] <- "cols"
+    names(args) <- nams
+    n_unnamed_args <- length(which(!nzchar(nams)))
+    all_args_unnamed <- n_unnamed_args == lengt
+  }
 
-  if (srow_null) srow <- 0 else srow <- srow - 1L
-  if (scol_null) scol <- 0 else scol <- col2int(scol) - 1L
+  # if 2 unnamed arguments, will be rows, cols.
+  if (n_unnamed_args == 2) {
+    # message("Assuming the first 2 unnamed arguments to be `rows`, `cols` resp.")
+    rows_pos <- which(nams == "")[1]
+    cols_pos <- which(nams == "")[2]
+    nams[c(rows_pos, cols_pos)] <- c("rows", "cols")
+    names(args) <- nams
+    n_unnamed_args <- length(which(!nzchar(nams)))
+    all_args_unnamed <- n_unnamed_args == lengt
+  }
 
-  x <- args[[1]]
-  exp_name <- inherits(x, "data.frame") || inherits(x, "matrix")
+  has_some_unnamed_args <- any(!nzchar(nams))
+  if (has_some_unnamed_args) {
+    stop("Internal error, all arguments should be named after this point.")
+  }
 
-  # wb_dims(rows, cols)
-  if (length(args) >= 2 && !exp_name) {
-    rows <- 1L
-    cols <- 2L
+  x_present <- "x" %in% nams
+  cond_acceptable_lengt_1 <- x_present || !is.null(args$from_row) || !is.null(args$from_col)
 
-    # wb_dims(rows = rows, cols = cols)
-    sel <- pmatch(nams, c("rows", "cols"))
-    valid <- length(sel[!is.na(sel)])
-    if (valid == 2) {
-      rows <- sel[rows]
-      cols <- sel[cols]
-    } else if (valid == 1) {
-      stop("found only one cols/rows argument")
+  if (lengt == 1 && !cond_acceptable_lengt_1) {
+    # Providing a single argument acceptable is only  `x`
+    sentence_unnamed <- ifelse(all_args_unnamed, "unnamed ", " ")
+    stop(
+      "Supplying a single", sentence_unnamed, "argument to `wb_dims()` is not supported.",
+      "\n",
+      "use any of `x`, `from_row` `from_col`. You can also use `rows` and `cols`, You can also use `dims = NULL`"
```

Comment:
`wb_dims(rows = 1)` this throws, even though `rows` is supplied. In this case `cols = 1` could be assumed and vice versa

Reply:
It gave the wrong result in the previous version...
I think that the user should really use
```r
wb_dims(from_row = 1)
```

```r
wb_dims()
wb_dims(NULL)
wb_dims(1, 1)
wb_dims(1, 2)
wb_dims(cols =1, 2)
```
are all acceptable alternatives

Reply:
OK, fine by me, as stated, we don't have to make this more complicated than it already is.